File: hackerrank/hr_counting-valleys.py
def countingValleys(steps, path):
    if steps == 0:
        return 0
    l = 0
    arr = []
    valleys = 0
    for x in path:
        if x == "U":
            l+=1
        else:
            l-=1
        if l == 0 and arr[-1] < 0:
            valleys+=1
        arr.append(l)
    return valleys



countingValleys(10,"UDUUUDUDDD")

File: hackerrank/hr_counting-valleys.cpp
#include <iostream>
#include <vector>

using namespace std;

int solve(int steps, string path){	
	vector<int> arr = {};
	int l = 0;
	int valleys = 0;
	for( int i = 0; i < path.length(); i++){
		char x = path.at(i);
		char u = 'U';
		if (x == u){
			l++;
		} else {
			l--;
		}
		
		if ( l == 0 && arr.at(arr.size()-1) < 0){
			cout << "FOUND VALLEY" << endl;
			valleys++;
		}
		arr.push_back(l);
		cout << x << endl;
	};

	return valleys;
}

int main(){
	int res = solve(8 ,"UDDDUDUU");
	cout << res << "\n";
	return 0 ;
}
